src/main/scala/com/microsoft/ml/spark/lightgbm/dataset/DatasetAggregator.scala

```diff
@@ -0,0 +1,488 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.ml.spark.lightgbm.dataset
+
+import com.microsoft.ml.spark.lightgbm.ColumnParams
+import com.microsoft.ml.spark.lightgbm.swig.{BaseSwigArray, ChunkedArray, DoubleChunkedArray, DoubleSwigArray,
+  FloatChunkedArray, FloatSwigArray, IntChunkedArray, IntSwigArray}
+import org.apache.spark.sql.Row
+
+import scala.collection.mutable.ListBuffer
+
+object ChunkedArrayUtils {
+  def copyChunkedArray[T](chunkedArray: ChunkedArray[T],
+                          mainArray: BaseSwigArray[T],
+                          threadRowStartIndex: Long,
+                          chunkSize: Long): Unit = {
+    // Copy in parallel on each thread
+    // First copy full chunks
+    val chunkCount = chunkedArray.getChunksCount() - 1
+    for (chunk <- 0L until chunkCount) {
+      for (in_chunk_idx <- 0L until chunkSize) {
+        mainArray.setItem(threadRowStartIndex + chunk * chunkSize + in_chunk_idx,
+          chunkedArray.getItem(chunk, in_chunk_idx, (-1).asInstanceOf[T]))
+      }
+    }
+    // Next copy filled values from last chunk only
+    val lastChunkCount = chunkedArray.getLastChunkAddCount()
+    for (last_chunk_idx <- 0L until lastChunkCount) {
+      mainArray.setItem(threadRowStartIndex + chunkCount * chunkSize + last_chunk_idx,
+        chunkedArray.getItem(chunkCount, last_chunk_idx, (-1).asInstanceOf[T]))
+    }
+  }
+}
+
+trait AggregatedColumns
```

Comment:
I used this to fit the prep/merge interface Mark wanted me to use... I would have to change the prep/merge interface to remove this trait.

Reply:
you can find the trait here:

https://github.com/Azure/mmlspark/blob/master/core/src/main/scala/com/microsoft/ml/spark/stages/PartitionConsolidator.scala#L140
```
trait LocalAggregator[T] {
  def prep(iter: Iterator[Row]): T

  def merge(ts: Seq[T]): T
}
```
I don't think it quite makes sense for me though because I'm really doing something more like:
```
trait LocalAggregator[TChunk, TAgg] {
  def prep(iter: Iterator[Row]): TChunk

  def merge(ts: TChunk): TAgg
}
```
maybe Mark might have some ideas here, but I tried to fit into the constraints of LocalAggregator and the easiest way to do this was to create the common trait AggregatedColumns